### ISSUE TITLE
[P4] Executable inter-host pipeline planning, admission, and recovery

### ISSUE THREAD

**Issue by ewindisch** [hyprstream/hyprstream#325]
Part of #310. Plan: `docs/plans/2026-06-18-multi-gpu-multi-host-inference-spike.md`. (M-FAIL/M-BUBBLE/M-2C-SCALE)

Multi-hop failure/recovery (stage death mid-request, partial KV, drain, HOL blocking). Confirm
continuous batching exists or build it; quantify bubble fraction `(S-1)/(m+S-1)` and required
microbatch depth. Capacity-aware stage planner from `config.json` (asymmetric VRAM). Envelope:
LAN-only, minimize boundaries; for multi-TB throughput serving, not single-stream latency.

**Comment by ewindisch** [comment]
Code-level recon at `origin/main` / `69eda14916c0870c16f101eb4147f637e09f2043`:

The repository has the right **pipeline-shaped primitives**, but neither production 2b nor 2c is
complete.

* `DevicePool` explicitly chooses one thread driving tensors on multiple devices
  (`crates/hyprstream/src/runtime/device_pool.rs:13-28`), and inference is constructed inside a
  current-thread Tokio runtime/`LocalSet` on a dedicated OS thread
  (`crates/hyprstream-rpc/src/transport/iroh_rpc.rs:387-437`,
  `crates/hyprstream/src/services/inference.rs:2507-2568`). This answers #311's architecture
  choice.
* The factory constructs a multi-device Llama/Qwen model
  (`crates/hyprstream/src/runtime/model_factory.rs:790-814`), but production cached inference calls
  `model.forward_with_cache(...)` (`crates/hyprstream/src/runtime/torch_engine.rs:1091-1098`).
  Llama's corresponding loop has no device-boundary move
  (`crates/hyprstream/src/runtime/architectures/llama.rs:2296-2339`). The intended
  `hidden.to_device(target)` exists only in `forward_layers_inner` (`:2632-2644`); an
  `rg -n "forward_layers\\(" crates/hyprstream/src --glob '*.rs'` search found only architecture
  tests and the unused training helper, not a production inference caller.
* The loader materializes all shard tensors on the primary device before constructing the model
  (`crates/hyprstream/src/runtime/model_factory.rs:252-284`, `:425-499`), so it cannot yet load a
  model larger than one GPU. Qwen3.5 explicitly falls back to one device (`:702-711`).
* There is no 2a engine pool. `DevicePool` calls N engines a future use
  (`crates/hyprstream/src/runtime/device_pool.rs:8-11`, `:185-199`); a source search for a
  `Vec<TorchEngine>`/per-device engine construction found none.
* Continuous batching is built as a Llama/TorchEngine primitive
  (`crates/hyprstream/src/runtime/architectures/mod.rs:258-279`,
  `crates/hyprstream/src/runtime/torch_engine.rs:1151-1184`) but is not wired:
  `rg -n "forward_batched_step\\(" crates/hyprstream/src --glob '*.rs'` found only its definition.

Therefore #311 is **partially stale**: the one-thread/N-device design is settled, but its
real-device feasibility gate is not. The existing GPU E2E deliberately stays single-device
(`crates/hyprstream/tests/e2e_inference.rs:721-724`), while the two-GPU-shaped test only constructs
device enums (`:610-631`). It should not close as “proved” until a real two-device load, prefill,
cached decode, RoPE reuse, and teardown pass exists.

For #324, the local contract to replace is a contiguous stage range where stage 0 embeds, the final
stage performs norm/head, and the only boundary tensor is `hidden [B,Q,H]`; `start_pos` is scalar
metadata, position IDs are reconstructed, and KV/SSM state remains stage-local
(`crates/hyprstream/src/runtime/architectures/mod.rs:376-413`). Payload size is
`B*Q*H*sizeof(dtype)`: at BF16 and H=4096, B1/Q8192 prefill is 64 MiB, B1/Q1 decode is 8 KiB, and
B16/Q1 decode is 128 KiB per boundary.

#324 needs to own an executable stage link, not just tensor serialization: model/plan identity,
request/session/epoch/microbatch/sequence identity, stage/layer boundary, prefill-vs-decode and
`start_pos`, dtype/shape/length, chunk layout/digest, byte-based receiver credits, exact
reassembly/order, cancellation, and authenticated terminal state. The existing MoQ path has
AEAD/HMAC/provenance support (`crates/hyprstream-rpc/src/moq_stream.rs:593-623`, `:733-805`,
`:1885-1928`), but its producer constructs a whole encoded frame and does not await downstream
credits (`:652-658`, `:694-826`). No explicit MoQ application payload cap was found; the 64 MiB
constant is the legacy ZMTP cap (`crates/hyprstream-rpc/src/zmtp_framing.rs:26-27`), so the exact
chunk ceiling remains **UNVERIFIED** and should be measured.

Failure should fence a pipeline epoch, reject stale completions, rebuild the chain, and replay the
prompt plus committed generated-token history to reconstruct stage-local caches. The durable
coordinator record needs model/plan/stage identities, prompt and token commits, sampling/RNG state,
tenant/delta revision, per-microbatch admission/completion, and the last client-visible output
commit.

#1078/#1222 are a useful spike, not merge-shaped production code. Keep contiguous ranges,
capacity-aware asymmetric placement/minimal boundaries
(`origin/pr/1078:crates/hyprstream/src/runtime/interhost_pipeline.rs:311-368`), bubble arithmetic as
a metric (`origin/pr/1222:crates/hyprstream/src/runtime/interhost_pipeline.rs:912-963`), and #1222's
typed identity/admission plus epoch/replay invariants (`:733-769`, `:993-1183`). Reshape placement
around exact indexed tensor bytes, runtime/KV/activation/reassembly memory, backend and link
capabilities, and executable stage descriptors. The current assignment has only host/range/two
byte counts (`origin/pr/1078:crates/hyprstream/src/runtime/interhost_pipeline.rs:287-302`).
`git grep -n InterhostPipelinePlanner origin/pr/1222 --
':!crates/hyprstream/src/runtime/interhost_pipeline.rs'` found only the `pub use`, not a caller.

Dependency order:

1. Wire the production device-aware forward and split-training helper (no open ticket; closed
   #314/#316 are incomplete by code).
2. Add true stage/subset loading and either wire Qwen3.5 stages or narrow the model contract (no
   ticket).
3. Resolve #311 with a real two-device gate.
4. Build 2a's N-engine replica pool (no ticket), then run #317.
5. Finish #329's scheduler around the existing batched kernel.
6. Add a network-addressable stage lifecycle/service (no ticket), then real remote replica
   resolution/dialing (no open ticket in the requested set).
7. Implement #324's activation protocol and stage integration.
8. Reshape/integrate #1078/#1222 under #325.
9. Run #323 cross-host/cross-vendor validation.
10. Add #327 inter-host backward communication.
11. Add tensor/expert parallelism only if a single layer cannot fit one device/host (conditional,
    no ticket); no collectives were found by
    `git grep -n -E 'all_reduce|tensor_parallel|world_size|nccl' HEAD --
    crates/hyprstream/src/runtime`.

#317 is diagnostic-runnable now but not acceptance-runnable: the current path will likely expose
the missing production copy, while no numerical-parity harness or 2a pool exists. After the
forward/loading fixes it needs deterministic single-vs-split full-logit and cached-decode
comparison, explicit proof both GPUs executed, repeated teardown, peak HBM, prefill/TTFT/decode,
boundary-copy latency, utilization, and P2P measurements.

#323 is not end-to-end runnable because no stage service/link exists. This week it can still
measure generic Iroh/MoQ RTT/throughput at 8 KiB, 128 KiB, and chunked 64 MiB; ROCm/CUDA
GPU↔CPU+codec+crypto costs; independent same-model backend drift; and synthetic stream faults.
The external MI210 checkout/runtime/topology/model inventory is **UNVERIFIED**.

Focused validation, all through `buildq -- cargo ...`: `device_pool --lib` **14 passed**;
`pipeline --lib` **20 passed**; `--test e2e_inference e6_layer` **5 passed**. These validate
planning and CPU/same-device equivalence, not a real cross-device boundary.

Full report:
`/home/birdetta/projects/cyberdione/.fleet-coord/mhi/REPORT.md`

**Comment by ewindisch** [comment]
**Rescoped (Epic B): executable planner + admission + recovery** (integration, not speculation). Remove "confirm continuous batching exists" (#329 owns the scheduler).

Planner consumes executable stage capabilities: model/stage ABI, OID, architecture, dtype, exact indexed weight bytes, backend/vendor, usable VRAM after runtime/KV, context/batch limits, compute rate, link latency/bandwidth/credit limits. Output = launchable chain (admitted identity/reach, exact range + first/last roles, shard selection, activation codec/chunk/credit contract, replicas, recovery policy).

Recovery persists/reconstructs prompt + committed token history, sampling/RNG or committed sampled-token log, tenant/delta revision, pipeline epoch, per-microbatch state, last client-visible output commit; fences stale completions, discards partial reassembly, reloads/replays/resumes without duplicate/skipped output. Deps: #314, #1236, #1237, #324, #329. **#1078/#1222 are extraction sources, not the implementation to merge.** — Epic B ([roadmap](https://claude.ai/code/artifact/8dad534b-c405-4d24-9637-f1781591bab4))
